File: lib/analytics/rfm.ts
/**
 * RFM анализ — сегментация клиентов по трём метрикам:
 *   R (Recency)   — насколько недавно была покупка (меньше дней = лучше)
 *   F (Frequency) — сколько покупок сделал клиент
 *   M (Monetary)  — суммарная выручка от клиента
 *
 * Каждой метрике присваивается балл 1–5 по квинтилям.
 * Итоговый код — RFM (например, 555 = VIP, 111 = потерянные).
 */

export type RfmTransaction = {
  customerId: string;
  customerName?: string;
  date: string | Date;
  amount: number;
};

export type RfmCustomer = {
  id: string;
  name: string;
  recencyDays: number;  // дни с последней покупки
  frequency: number;    // число покупок
  monetary: number;     // суммарная выручка
};

export type RfmScored = RfmCustomer & {
  rScore: number; // 1..5 (5 = недавно покупал)
  fScore: number; // 1..5 (5 = много покупок)
  mScore: number; // 1..5 (5 = большой чек)
  rfm: string;    // "555"
  segment: RfmSegment;
};

export type RfmSegment =
  | 'Champions'         // VIP: 555, 554, 544, 545
  | 'Loyal'             // лояльные
  | 'Potential Loyal'   // потенциально лояльные
  | 'New'               // новые
  | 'Promising'         // перспективные
  | 'Need Attention'    // требуют внимания
  | 'At Risk'           // в зоне риска
  | 'Hibernating'       // спящие
  | 'Lost';             // потерянные

export type RfmOptions = {
  referenceDate?: Date; // дата отсчёта recency (по умолчанию сегодня)
};

export function aggregateTransactions(
  txs: RfmTransaction[],
  refDate: Date,
): RfmCustomer[] {
  const map = new Map<string, RfmCustomer>();
  for (const t of txs) {
    const id = t.customerId;
    const d = t.date instanceof Date ? t.date : new Date(t.date);
    if (!Number.isFinite(d.getTime())) continue;
    const ageMs = refDate.getTime() - d.getTime();
    const ageDays = Math.max(0, Math.floor(ageMs / (1000 * 60 * 60 * 24)));
    const prev = map.get(id);
    if (!prev) {
      map.set(id, {
        id,
        name: t.customerName ?? id,
        recencyDays: ageDays,
        frequency: 1,
        monetary: t.amount,
      });
    } else {
      prev.recencyDays = Math.min(prev.recencyDays, ageDays);
      prev.frequency += 1;
      prev.monetary += t.amount;
      if (t.customerName) prev.name = t.customerName;
    }
  }
  return [...map.values()];
}

function quintileScore(value: number, sortedAsc: number[], invert = false): number {
  if (sortedAsc.length === 0) return 1;
  if (sortedAsc.length < 5) {
    const m = Math.max(...sortedAsc);
    const min = Math.min(...sortedAsc);
    if (m === min) return 3;
    const ratio = (value - min) / (m - min);
    const s = Math.min(5, Math.max(1, Math.ceil(ratio * 5)));
    return invert ? 6 - s : s;
  }
  const n = sortedAsc.length;
  const idx = sortedAsc.findIndex((v) => v >= value);
  const rank = idx === -1 ? n : idx + 1;
  const q = Math.min(5, Math.max(1, Math.ceil((rank / n) * 5)));
  return invert ? 6 - q : q;
}

export function scoreCustomers(customers: RfmCustomer[]): RfmScored[] {
  if (customers.length === 0) return [];
  const rSorted = [...customers].map((c) => c.recencyDays).sort((a, b) => a - b);
  const fSorted = [...customers].map((c) => c.frequency).sort((a, b) => a - b);
  const mSorted = [...customers].map((c) => c.monetary).sort((a, b) => a - b);

  return customers.map((c) => {
    // Recency меньше = лучше, поэтому инвертируем
    const rScore = quintileScore(c.recencyDays, rSorted, true);
    const fScore = quintileScore(c.frequency, fSorted);
    const mScore = quintileScore(c.monetary, mSorted);
    const rfm = `${rScore}${fScore}${mScore}`;
    return {
      ...c,
      rScore,
      fScore,
      mScore,
      rfm,
      segment: segmentOf(rScore, fScore, mScore),
    };
  });
}

function segmentOf(r: number, f: number, m: number): RfmSegment {
  const fm = (f + m) / 2;
  if (r >= 4 && fm >= 4) return 'Champions';
  if (r >= 3 && fm >= 3) return 'Loyal';
  if (r >= 4 && fm >= 2) return 'Potential Loyal';
  if (r === 5 && f === 1) return 'New';
  if (r >= 3 && fm <= 2) return 'Promising';
  if (r === 3 && fm >= 3) return 'Need Attention';
  if (r <= 2 && fm >= 3) return 'At Risk';
  if (r <= 2 && fm === 2) return 'Hibernating';
  return 'Lost';
}

export function buildRfmReport(
  txs: RfmTransaction[],
  opts: RfmOptions = {},
): RfmScored[] {
  const refDate = opts.referenceDate ?? new Date();
  const customers = aggregateTransactions(txs, refDate);
  return scoreCustomers(customers).sort((a, b) => b.monetary - a.monetary);
}

export function summarizeRfm(scored: RfmScored[]) {
  const segments = new Map<RfmSegment, { count: number; revenue: number }>();
  for (const s of scored) {
    const prev = segments.get(s.segment) ?? { count: 0, revenue: 0 };
    prev.count += 1;
    prev.revenue += s.monetary;
    segments.set(s.segment, prev);
  }
  return segments;
}

File: backend/app/analytics/rfm.py
"""
RFM анализ — порт widget/lib/analytics/rfm.ts.

Сегментация клиентов по трём метрикам:
  R (Recency)   — дней с последней покупки (меньше = лучше)
  F (Frequency) — число покупок
  M (Monetary)  — суммарная выручка

Каждой метрике — балл 1-5 по квинтилям. Сегмент определяется по матрице
(r, f, m). Логика и пороги сегментов один-в-один с TS-версией.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime
from typing import Literal


RfmSegment = Literal[
    "Champions",
    "Loyal",
    "Potential Loyal",
    "New",
    "Promising",
    "Need Attention",
    "At Risk",
    "Hibernating",
    "Lost",
]


@dataclass(frozen=True)
class RfmTransaction:
    customer_id: str
    customer_name: str | None
    date: datetime
    amount: float


@dataclass
class RfmCustomer:
    id: str
    name: str
    recency_days: int     # дни с последней покупки
    frequency: int        # число покупок
    monetary: float       # суммарная выручка


@dataclass
class RfmScored:
    id: str
    name: str
    recency_days: int
    frequency: int
    monetary: float
    r_score: int          # 1..5 (5 = недавно покупал)
    f_score: int          # 1..5 (5 = много покупок)
    m_score: int          # 1..5 (5 = большой чек)
    rfm: str              # "555"
    segment: RfmSegment


def aggregate_transactions(
    txs: list[RfmTransaction],
    ref_date: datetime,
) -> list[RfmCustomer]:
    """Группируем транзакции по customer_id, агрегируем R/F/M."""
    by_id: dict[str, RfmCustomer] = {}
    for t in txs:
        if not math.isfinite(t.date.timestamp()):
            continue
        age_ms = ref_date.timestamp() - t.date.timestamp()
        age_days = max(0, int(age_ms // 86400))
        prev = by_id.get(t.customer_id)
        if prev is None:
            by_id[t.customer_id] = RfmCustomer(
                id=t.customer_id,
                name=t.customer_name or t.customer_id,
                recency_days=age_days,
                frequency=1,
                monetary=t.amount,
            )
        else:
            prev.recency_days = min(prev.recency_days, age_days)
            prev.frequency += 1
            prev.monetary += t.amount
            if t.customer_name:
                prev.name = t.customer_name
    return list(by_id.values())


def _quintile_score(value: float, sorted_asc: list[float], invert: bool = False) -> int:
    """
    Балл 1..5 по квинтилю. Алгоритм идентичен TS-версии:
    - Если значений < 5 — линейная нормализация min..max в 1..5.
    - Если значений >= 5 — позиция в отсортированном массиве делится на 5.
    invert=True для recency (меньше = лучше).
    """
    if not sorted_asc:
        return 1
    if len(sorted_asc) < 5:
        m = max(sorted_asc)
        mn = min(sorted_asc)
        if m == mn:
            return 3
        ratio = (value - mn) / (m - mn)
        s = min(5, max(1, math.ceil(ratio * 5)))
        return 6 - s if invert else s
    n = len(sorted_asc)
    # findIndex эквивалент: первый индекс где sorted_asc[i] >= value
    idx = next((i for i, v in enumerate(sorted_asc) if v >= value), -1)
    rank = n if idx == -1 else idx + 1
    q = min(5, max(1, math.ceil((rank / n) * 5)))
    return 6 - q if invert else q


def _segment_of(r: int, f: int, m: int) -> RfmSegment:
    """Матрица сегментов из TS-версии (порядок важен — first-match)."""
    fm = (f + m) / 2
    if r >= 4 and fm >= 4:
        return "Champions"
    if r >= 3 and fm >= 3:
        return "Loyal"
    if r >= 4 and fm >= 2:
        return "Potential Loyal"
    if r == 5 and f == 1:
        return "New"
    if r >= 3 and fm <= 2:
        return "Promising"
    if r == 3 and fm >= 3:
        return "Need Attention"
    if r <= 2 and fm >= 3:
        return "At Risk"
    if r <= 2 and fm == 2:
        return "Hibernating"
    return "Lost"


def score_customers(customers: list[RfmCustomer]) -> list[RfmScored]:
    if not customers:
        return []
    r_sorted = sorted(c.recency_days for c in customers)
    f_sorted = sorted(c.frequency for c in customers)
    m_sorted = sorted(c.monetary for c in customers)

    result: list[RfmScored] = []
    for c in customers:
        r_score = _quintile_score(c.recency_days, r_sorted, invert=True)
        f_score = _quintile_score(c.frequency, f_sorted)
        m_score = _quintile_score(c.monetary, m_sorted)
        result.append(RfmScored(
            id=c.id, name=c.name,
            recency_days=c.recency_days,
            frequency=c.frequency,
            monetary=c.monetary,
            r_score=r_score,
            f_score=f_score,
            m_score=m_score,
            rfm=f"{r_score}{f_score}{m_score}",
            segment=_segment_of(r_score, f_score, m_score),
        ))
    return result


def build_rfm_report(
    txs: list[RfmTransaction],
    reference_date: datetime | None = None,
) -> list[RfmScored]:
    ref = reference_date or datetime.now()
    customers = aggregate_transactions(txs, ref)
    scored = score_customers(customers)
    # Сортировка по monetary desc — самые ценные клиенты первыми
    scored.sort(key=lambda s: s.monetary, reverse=True)
    return scored


def summarize_rfm(scored: list[RfmScored]) -> dict[str, dict[str, float]]:
    """Эквивалент summarizeRfm: {segment: {count, revenue}}."""
    result: dict[str, dict[str, float]] = {}
    for s in scored:
        bucket = result.setdefault(s.segment, {"count": 0, "revenue": 0.0})
        bucket["count"] += 1
        bucket["revenue"] += s.monetary
    return result
